packages/typescript-estree/src/useProgramFromProjectService.ts

```diff
@@ -42,70 +47,107 @@ const updateExtraFileExtensions = (
     /* scriptKind */ undefined,
     parseSettings.tsconfigRootDir,
   );
 
-  log('Opened project service file: %o', opened);
+  log(
+    'Project service type information enabled; checking for file path match on: %o',
+    serviceSettings.allowDefaultProject,
+  );
 
-  if (hasFullTypeInformation) {
-    log(
-      'Project service type information enabled; checking for file path match on: %o',
-      allowDefaultProject,
-    );
-    const isDefaultProjectAllowedPath = filePathMatchedBy(
-      parseSettings.filePath,
-      allowDefaultProject,
-    );
+  log(
+    'Default project allowed path: %s, based on config file: %s',
+    isDefaultProjectAllowed,
+    opened.configFileName,
+  );
 
-    log(
-      'Default project allowed path: %s, based on config file: %s',
-      isDefaultProjectAllowedPath,
-      opened.configFileName,
+  if (opened.configFileName) {
+    if (isDefaultProjectAllowed) {
+      throw new Error(
+        `${parseSettings.filePath} was included by allowDefaultProject but also was found in the project service. Consider removing it from allowDefaultProject.`,
+      );
+    }
+  } else if (!isDefaultProjectAllowed) {
+    throw new Error(
+      `${parseSettings.filePath} was not found by the project service. Consider either including it in the tsconfig.json or including it in allowDefaultProject.`,
     );
+  }
+
+  // No a configFileName indicates this file wasn't included in a TSConfig.
+  // That means it must get its type information from the default project.
+  if (!opened.configFileName) {
+    defaultProjectMatchedFiles.add(filePathAbsolute);
+    if (
+      defaultProjectMatchedFiles.size >
+      serviceSettings.maximumDefaultProjectFileMatchCount
+    ) {
+      const filePrintLimit = 20;
+      const filesToPrint = Array.from(defaultProjectMatchedFiles).slice(
+        0,
+        filePrintLimit,
+      );
+      const truncatedFileCount =
+        defaultProjectMatchedFiles.size - filesToPrint.length;
 
-    if (opened.configFileName) {
-      if (isDefaultProjectAllowedPath) {
-        throw new Error(
-          `${parseSettings.filePath} was included by allowDefaultProject but also was found in the project service. Consider removing it from allowDefaultProject.`,
-        );
-      }
-    } else if (!isDefaultProjectAllowedPath) {
       throw new Error(
-        `${parseSettings.filePath} was not found by the project service. Consider either including it in the tsconfig.json or including it in allowDefaultProject.`,
+        `Too many files (>${serviceSettings.maximumDefaultProjectFileMatchCount}) have matched the default project.${DEFAULT_PROJECT_FILES_ERROR_EXPLANATION}
+Matching files:
+${filesToPrint.map(file => `- ${file}`).join('\n')}
+${truncatedFileCount ? `...and ${truncatedFileCount} more files\n` : ''}
+If you absolutely need more files included, set parserOptions.projectService.maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING to a larger value.
+`,
```

Comment:
Note that I've moved this _"Too many files ..."_ notice to _inside_ the `if (opened.configFileName)` in this PR. That should stop the notice from getting logged on all files!